[SPARK-12297][SQL] Hive compatibility for Parquet Timestamps

squito opened the pull request:
## What changes were proposed in this pull request?

This change allows timestamps in parquet-based hive table to behave as a "floating time", without a timezone, as timestamps are for other file formats.  If the storage timezone is the same as the session timezone, this conversion is a no-op.  When data is read from a hive table, the table property is *always* respected.  This allows spark to not change behavior when reading old data, but read newly written data correctly (whatever the source of the data is).

Spark inherited the original behavior from Hive, but Hive is also updating behavior to use the same  scheme in HIVE-12767 / HIVE-16231.

The default for Spark remains unchanged; created tables do not include the new table property.

This will only apply to hive tables; nothing is added to parquet metadata to indicate the timezone, so data that is read or written directly from parquet files will never have any conversions applied.

## How was this patch tested?

Added a unit test which creates tables, reads and writes data, under a variety of permutations (different storage timezones, different session timezones, vectorized reading on and off).

squito committed:
very basic test for adjusting read parquet data

squito committed:
wip

squito committed:
working version for non-vectorized read -- lots of garbage too

squito committed:
working for vectorized reads -- not sure about all code paths

squito committed:
more tests for write path

squito committed:
expand tests; fix some metastore interaction; cleanup a lot of garbage

squito committed:
more cleanup

squito committed:
handle bad timezones; include unit test

squito committed:
write support; lots more unit tests

squito committed:
add tests for alter table

squito committed:
utc or gmt; cleanup

squito committed:
more cleanup

SparkQA commented:
**[Test build #72287 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72287/testReport)** for PR 16781 at commit [`223ce2c`](https://github.com/apache/spark/commit/223ce2c25b122707c64e4eda77a11bff71fd0cbe).

SparkQA commented:
**[Test build #72287 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72287/testReport)** for PR 16781 at commit [`223ce2c`](https://github.com/apache/spark/commit/223ce2c25b122707c64e4eda77a11bff71fd0cbe).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/72287/
Test FAILed.

squito committed:
fix compatibility

SparkQA commented:
**[Test build #72288 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72288/testReport)** for PR 16781 at commit [`5b49ae0`](https://github.com/apache/spark/commit/5b49ae026044b46f0899a9e792e2b71733c4cb8a).

SparkQA commented:
**[Test build #72288 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72288/testReport)** for PR 16781 at commit [`5b49ae0`](https://github.com/apache/spark/commit/5b49ae026044b46f0899a9e792e2b71733c4cb8a).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/72288/
Test FAILed.

squito committed:
Merge branch 'master' into SPARK-12297

squito committed:
Merge branch 'master' into SPARK-12297

squito committed:
wip

squito committed:
fix

squito committed:
fixes; passes tests now

squito committed:
Merge branch 'master' into SPARK-12297

squito committed:
fix merge

SparkQA commented:
**[Test build #74031 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74031/testReport)** for PR 16781 at commit [`2c8a228`](https://github.com/apache/spark/commit/2c8a22811f404c751841e9d9f2e8b22780d60f99).

SparkQA commented:
**[Test build #74031 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74031/testReport)** for PR 16781 at commit [`2c8a228`](https://github.com/apache/spark/commit/2c8a22811f404c751841e9d9f2e8b22780d60f99).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/74031/
Test FAILed.

squito committed:
fix

SparkQA commented:
**[Test build #74042 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74042/testReport)** for PR 16781 at commit [`f0b89fd`](https://github.com/apache/spark/commit/f0b89fdca05d61bff7f79047b5ac72125bfd107e).

SparkQA commented:
**[Test build #74042 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74042/testReport)** for PR 16781 at commit [`f0b89fd`](https://github.com/apache/spark/commit/f0b89fdca05d61bff7f79047b5ac72125bfd107e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/74042/
Test PASSed.

zivanfi commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedColumnReader.java:
```diff
@@ -89,11 +92,23 @@
+    // The conf is sometimes null in tests.
+    String tzString =
+        conf == null ? null : conf.get(ParquetFileFormat.PARQUET_TIMEZONE_TABLE_PROPERTY());
+    if (tzString == null || tzString == "") {
```
This is java code, not scala, you probably meant tzString.equals("") instead of tzString == ""

zivanfi replied:
Or even better, isEmpty().

zivanfi commented on sql/core/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:
```diff
@@ -674,6 +674,12 @@ object SQLConf {
 
+  val PARQUET_TABLE_INCLUDE_TIMEZONE =
+    buildConf("spark.sql.session.parquet.timeZone")
+      .doc("""Enables inclusion of parquet timezone property in newly created parquet tables""")
```
There should be a config option for writing "UTC" to the table property when creating tables, not for writing the local timezone.

zivanfi commented:
Please update the pull request description, because the one dated Feb 2 does not correspond to the fix any more.

zivanfi commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/ParquetHiveCompatibilitySuite.scala:
```diff
@@ -137,8 +141,190 @@ class ParquetHiveCompatibilitySuite extends ParquetCompatibilityTest with TestHi
       "ARRAY<STRUCT<array_element: INT>>")
   }
+
+  test(s"SPARK-12297: Parquet Timestamp & Hive timezone") {
```
I think it would be better to have separate test cases for adjustments when reading, adjustments when writing and setting the table property when creating tables.

squito committed:
Merge branch 'master' into SPARK-12297

squito committed:
refactor the test

squito committed:
cleanup

SparkQA commented:
**[Test build #74459 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74459/testReport)** for PR 16781 at commit [`c242fb8`](https://github.com/apache/spark/commit/c242fb89e2861c81c44877a3acfded9073b9104b).

SparkQA commented:
**[Test build #74459 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74459/testReport)** for PR 16781 at commit [`c242fb8`](https://github.com/apache/spark/commit/c242fb89e2861c81c44877a3acfded9073b9104b).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/74459/
Test FAILed.

squito committed:
reset timezone property after tests; make tests more granular

SparkQA commented:
**[Test build #74542 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74542/testReport)** for PR 16781 at commit [`c87a573`](https://github.com/apache/spark/commit/c87a57323ac1264ed4528a8cd2e26d342ca117f6).

SparkQA commented:
**[Test build #74542 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74542/testReport)** for PR 16781 at commit [`c87a573`](https://github.com/apache/spark/commit/c87a57323ac1264ed4528a8cd2e26d342ca117f6).
 * This patch passes all tests.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/74542/
Test PASSed.

squito committed:
Merge branch 'master' into SPARK-12297

SparkQA commented:
**[Test build #74550 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74550/testReport)** for PR 16781 at commit [`db7e514`](https://github.com/apache/spark/commit/db7e5147c09d9f773e630982d02709ca06b3b638).

SparkQA commented:
**[Test build #74550 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74550/testReport)** for PR 16781 at commit [`db7e514`](https://github.com/apache/spark/commit/db7e5147c09d9f773e630982d02709ca06b3b638).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/74550/
Test PASSed.

squito committed:
separate tests for reading & writing

SparkQA commented:
**[Test build #74609 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74609/testReport)** for PR 16781 at commit [`f4dca27`](https://github.com/apache/spark/commit/f4dca27e301c1ce94e7175c6a852a4e94cea0555).

SparkQA commented:
**[Test build #74609 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74609/testReport)** for PR 16781 at commit [`f4dca27`](https://github.com/apache/spark/commit/f4dca27e301c1ce94e7175c6a852a4e94cea0555).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/74609/
Test FAILed.

squito committed:
Merge branch 'master' into SPARK-12297

squito committed:
fix merge

SparkQA commented:
**[Test build #74611 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74611/testReport)** for PR 16781 at commit [`d951443`](https://github.com/apache/spark/commit/d951443d7dddd187fe119f01cb9ee16459f4a346).

SparkQA commented:
**[Test build #74611 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74611/testReport)** for PR 16781 at commit [`d951443`](https://github.com/apache/spark/commit/d951443d7dddd187fe119f01cb9ee16459f4a346).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/74611/
Test PASSed.

squito committed:
Merge branch 'master' into SPARK-12297

SparkQA commented:
**[Test build #74688 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74688/testReport)** for PR 16781 at commit [`38e19cd`](https://github.com/apache/spark/commit/38e19cdd497992fb063cb39d1d65bde1622553e4).

SparkQA commented:
**[Test build #74688 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74688/testReport)** for PR 16781 at commit [`38e19cd`](https://github.com/apache/spark/commit/38e19cdd497992fb063cb39d1d65bde1622553e4).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/74688/
Test PASSed.

squito commented:
pinging some potential reviewers:
@liancheng @yhuai for prior work in hive compatibility
@ueshin for work on timezone support

Note that the timezone support from SPARK-18350 is to some extent at odds with this.   Its not a total conflict, because that is addressing how spark stores data in its own json & csv format (among others).  But this is about making sure that *hive* tables are consistent between different formats.  I am somewhat puzzled by those changes, as they seem to be changing the meaning of the "TIMESTAMP" datatype to be more like "TIMESTAMP WITH TIMEZONE" in other sql engines, but I'll also admit I'm out of my expertise here.

This is not the area of code I am most familiar with, happy to get feedback.  One thing in particular I haven't addressed is predicate pushdown -- in particular, I'm not sure how to *test* predicate pushdown.  I can poke at that some more, but I'd appreciate any guidance.

ueshin commented:
I checked HIVE-12767 and reviewed this pr roughly.
And I collected my thoughts about the desired behavior of this issue, please correct me if I'm wrong.

when creating table:

- if `SQLConf.PARQUET_TABLE_INCLUDE_TIMEZONE` == true
  - if a property `PARQUET_TIMEZONE_TABLE_PROPERTY` exists
    - include the table property `PARQUET_TIMEZONE_TABLE_PROPERTY`
    - use the `PARQUET_TIMEZONE_TABLE_PROPERTY` value
  - else
    - include the table property `PARQUET_TIMEZONE_TABLE_PROPERTY`
    - use session local timezone as the default `PARQUET_TIMEZONE_TABLE_PROPERTY` value
- else
  - if a property `PARQUET_TIMEZONE_TABLE_PROPERTY` exists
    - include the table property `PARQUET_TIMEZONE_TABLE_PROPERTY`
    - use the `PARQUET_TIMEZONE_TABLE_PROPERTY` value
  - else
    - don't include table property `PARQUET_TIMEZONE_TABLE_PROPERTY`

when writing/reading data:

- if a table property `PARQUET_TIMEZONE_TABLE_PROPERTY` exists
  - use the `PARQUET_TIMEZONE_TABLE_PROPERTY` value to adjust timezone
- else
  - don't adjust timezone

Timezone related expressions respect session local timezone now, so we should also use session local timezone as the default value of `PARQUET_TIMEZONE_TABLE_PROPERTY` instead of system timezone, i.e. use `sparkSession.sessionState.conf.sessionLocalTimeZone` instead of `TimeZone.getDefault()`.

squito committed:
Merge branch 'master' into SPARK-12297

squito committed:
cleanup

squito commented:
@ueshin thanks for taking a look.  Yes, I think your summary is exactly correct.

Your suggestion of using the session timezone makes a lot of sense.  I can update the pr accordingly.  But I had another thought I wanted to run by you first -- perhaps in the initial version, we should *never* set the table property automatically?  Just leave it to users to set the property explicitly (in "create table" or "alter table")?

SparkQA commented:
**[Test build #75101 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75101/testReport)** for PR 16781 at commit [`39f506c`](https://github.com/apache/spark/commit/39f506c81187b466809ebca21b7d1bc4c1e82f7a).

ueshin commented:
Do you mean the default `SQLConf.PARQUET_TABLE_INCLUDE_TIMEZONE` should be `false` for now?
If so, I agree with it.

SparkQA commented:
**[Test build #75101 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75101/testReport)** for PR 16781 at commit [`39f506c`](https://github.com/apache/spark/commit/39f506c81187b466809ebca21b7d1bc4c1e82f7a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/75101/
Test PASSed.

squito commented:
> Do you mean the default SQLConf.PARQUET_TABLE_INCLUDE_TIMEZONE should be false for now?
If so, I agree with it.

I was proposing something more drastic -- eliminating that conf entirely.  If you want your table to have the timezone property set, rather than using that conf, you include `TBLPROPERTIES ("parquet.mr.int96.write.zone"="[some timezone]")` in your ddl (either "create table" or "alter table")

ueshin commented:
Ah, I see, it looks good to me for now.

squito committed:
Merge branch 'master' into SPARK-12297

squito committed:
remove config for setting table time zone automatically

squito committed:
fixup

SparkQA commented:
**[Test build #75177 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75177/testReport)** for PR 16781 at commit [`a96806f`](https://github.com/apache/spark/commit/a96806fa4011e28dcc875d0fb519501013b91a1b).

SparkQA commented:
**[Test build #75177 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75177/testReport)** for PR 16781 at commit [`a96806f`](https://github.com/apache/spark/commit/a96806fa4011e28dcc875d0fb519501013b91a1b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/75177/
Test PASSed.

squito commented:
@ueshin I've updated this to remove the conf entirely.  So updating your previous description, the new behavior can be described as:

when creating table:

* if a property `PARQUET_TIMEZONE_TABLE_PROPERTY` exists
    * include the table property `PARQUET_TIMEZONE_TABLE_PROPERTY`
    * use the `PARQUET_TIMEZONE_TABLE_PROPERTY` value
* else
    * don't include table property `PARQUET_TIMEZONE_TABLE_PROPERTY`

when writing/reading data:

* if a table property `PARQUET_TIMEZONE_TABLE_PROPERTY` exists
    * use the `PARQUET_TIMEZONE_TABLE_PROPERTY` value to adjust timezone
* else
    * don't adjust timezone

ueshin commented:
I looked over this pr and I noticed that you're using the `TimeZone.getDefault()` as the "local" timezone.
Since the timestamp values are represented as "the number of micros since epoch" internally, we should use `GMT` as the "local" timezone.

ueshin commented:
Btw, I noticed that if we decided to use `SQLConf.PARQUET_TABLE_INCLUDE_TIMEZONE`, the default `PARQUET_TIMEZONE_TABLE_PROPERTY` value should be `GMT` instead of session local timezone as Hive does.

squito commented:
thanks @ueshin ... I am going to chat w/ some folks involved in that hive patch, that was not my understanding conceptually of their patch.  I heard that there is a bug they need to fix so maybe its related.  I'll find out and update here appropriately.

squito committed:
predicate pushdown tests

squito committed:
minor cleanup

squito committed:
Merge branch 'master' into SPARK-12297

squito committed:
fix merge

squito committed:
swap conversion logic

squito committed:
update tests a bunch; tests session timezones, and tests are easier to reason about now

squito committed:
session timezones

squito committed:
cleanup

squito committed:
Merge branch 'master' into SPARK-12297

SparkQA commented:
**[Test build #75686 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75686/testReport)** for PR 16781 at commit [`75e8579`](https://github.com/apache/spark/commit/75e8579551c2f1ae7ae958853754fbbc5a589dd4).

SparkQA commented:
**[Test build #75686 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75686/testReport)** for PR 16781 at commit [`75e8579`](https://github.com/apache/spark/commit/75e8579551c2f1ae7ae958853754fbbc5a589dd4).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/75686/
Test PASSed.

squito commented:
@ueshin thanks for taking a look earlier, sorry it has taken me some time to update this.

Things to note since last time:

1) Hive has seen been updated in [HIVE-16231](https://issues.apache.org/jira/browse/HIVE-16231) to use the local timezone, not GMT, as the default for storing data.  Really, this is the change that should have been in HIVE-12767 -- otherwise you lose backwards compatibility with old datasets.

2) This PR now uses the session time zone, rather than local timezone.  There are tests to confirm that a mix of session timezone X storage timezone works correctly.

3) Predicate pushdown is handled.  I actually didn't need to change the behavior at all, since predicates are never pushed to int96 -- but there are some tests that confirm this.

I'm sure there is some minor cleanup that could be done, but overall I think this is ready now.  I'd appreciate if you take another look and any suggestions you can make.

ueshin commented:
@squito I'm sorry for late reply.
I was reviewing this pr again and current Hive behavior deeply.

It seems the both behaviors are the same when the table property exists, but this is a little complicated, so I want to confirm if the following behavior is correct or not by the simple example:

```scala

scala> spark.conf.set("spark.sql.session.timeZone", "JST")

scala> sql("create table foo (display string, ts timestamp) stored as parquet tblproperties (parquet.mr.int96.write.zone='JST')")
scala> sql("insert into foo values ('2017-01-01 00:00:00', cast('2017-01-01 00:00:00' as timestamp))")

scala> sql("select * from foo").show(truncate = false)
+-------------------+-------------------+
|display            |ts                 |
+-------------------+-------------------+
|2017-01-01 00:00:00|2017-01-01 00:00:00|
+-------------------+-------------------+


scala> sql("create table bar (display string, ts timestamp) stored as parquet")
scala> sql("insert into bar values ('2017-01-01 00:00:00', cast('2017-01-01 00:00:00' as timestamp))")

scala> sql("select * from bar").show(truncate = false)
+-------------------+-------------------+
|display            |ts                 |
+-------------------+-------------------+
|2017-01-01 00:00:00|2017-01-01 00:00:00|
+-------------------+-------------------+


scala> spark.conf.set("spark.sql.session.timeZone", "PST")


// ts is adjusted from JST to PST.
scala> sql("select * from foo").show(truncate = false)
+-------------------+-------------------+
|display            |ts                 |
+-------------------+-------------------+
|2017-01-01 00:00:00|2017-01-01 00:00:00|
+-------------------+-------------------+


// ts isn't adjusted.
scala> sql("select * from bar").show(truncate = false)
+-------------------+-------------------+
|display            |ts                 |
+-------------------+-------------------+
|2017-01-01 00:00:00|2016-12-31 07:00:00|
+-------------------+-------------------+
```

ueshin reviewed (commented):
And I added some comments for the current pr.

squito commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala:
```diff
@@ -498,6 +498,11 @@ object DateTimeUtils {
 
+  lazy val validTimezones = TimeZone.getAvailableIDs().toSet
+  def isValidTimezone(timezoneId: String): Boolean = {
+    validTimezones.contains(timezoneId)
```
Java does a case-sensitive check, which means Hive does too.  I don't think we want to write out a timezone w/ the wrong capitalization, and then have another tool throw an error.

```scala
scala> val tzId = "America/Los_Angeles"
tzId: String = America/Los_Angeles

scala> java.util.TimeZone.getTimeZone(tzId).getID()
res1: String = America/Los_Angeles

scala> java.util.TimeZone.getTimeZone(tzId.toLowerCase()).getID()
res2: String = GMT
```

https://github.com/apache/hive/blob/master/ql/src/java/org/apache/hadoop/hive/ql/io/parquet/timestamp/NanoTimeUtils.java#L167

We could try to auto-convert the user's timezone to the correct capitilazation, but do you think that is worth it?

ueshin replied:
Ah, I see. Let's keep current work.

squito commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRowConverter.scala:
```diff
@@ -261,18 +266,20 @@ private[parquet] class ParquetRowConverter(
       case TimestampType =>
         // TODO Implements `TIMESTAMP_MICROS` once parquet-mr has that.
+        // If the table has a timezone property, apply the correct conversions.  See SPARK-12297.
+        val sessionTz = TimeZone.getTimeZone(hadoopConf.get(SQLConf.SESSION_LOCAL_TIMEZONE.key))
```
good catch -- this would fail if you bypassed a hive table and read directly with `spark.read.parquet()`.  I'll be sure to add a test case to cover it as well.

squito commented:
@ueshin thanks for taking a look.  Yes, that understanding is correct.  Another way to think about it is to compare those same operations with different file formats, eg. textfile.  Those work more like parquet does *after* this patch.  I had that explanation in a comment on the jira -- I just updated the jira description to include it.

I'll address your comments, they also are making me take a closer look at a couple of things.  I should push an update tomorrow.

squito committed:
partial review feedback

squito committed:
better param names and docs

SparkQA commented:
**[Test build #75966 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75966/testReport)** for PR 16781 at commit [`44a8bbb`](https://github.com/apache/spark/commit/44a8bbb17484a61dd984fcb451e3b1be8c539e9f).

squito commented:
@ueshin I've pushed an update which addresses your comments.  I also realized that partitioned tables weren't handled correctly!  I fixed that as well.

squito commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/ParquetHiveCompatibilitySuite.scala:
```diff
@@ -141,4 +160,326 @@ class ParquetHiveCompatibilitySuite extends ParquetCompatibilityTest with TestHi
+          )
+          Seq(false, true).foreach { vectorized =>
+            withClue(s"vectorized = $vectorized;") {
+              spark.conf.set(SQLConf.PARQUET_VECTORIZED_READER_ENABLED.key, vectorized)
```
I was initially using `SQLTestUtils.withSQLConf`, but I discovered that it wasn't actually taking any effect. I dunno if that is because `TestHiveSingleton` does something strange, or maybe I'm doing something else weird in this test by creating many new spark sessions.  But I did that because it was the only way I could get the conf changes applied consistently.

Since I am creating new sessions, I don't think this has any risk of a failed test not cleaning and triggering failures in other tests outside of this suite.  But it still seems like I might be doing something wrong ...

ueshin commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetRowConverter.scala:
```diff
@@ -673,12 +679,35 @@ private[parquet] object ParquetRowConverter {
+    *                  the hour.
+   * @param storageTz the timezone which was used to store the timestamp.  This should come from the
+    *                  timestamp table property, or else assume its the same as the sessionTz
+   * @return
```
Can you also add descriptions for `@param binary` and `@return`?

ueshin commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/ParquetHiveCompatibilitySuite.scala:
```diff
@@ -141,4 +160,326 @@ class ParquetHiveCompatibilitySuite extends ParquetCompatibilityTest with TestHi
+  }
+
+  private def checkHasTz(table: String, tz: Option[String]): Unit = {
+    val tableMetadata = spark.sessionState.catalog.getTableMetadata(TableIdentifier(table))
```
Should explicitly pass `sparkSession` and use it here?

ueshin commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/ParquetHiveCompatibilitySuite.scala:
```diff
@@ -141,4 +160,326 @@ class ParquetHiveCompatibilitySuite extends ParquetCompatibilityTest with TestHi
+      val key = ParquetFileFormat.PARQUET_TIMEZONE_TABLE_PROPERTY
+      withTable(baseTable, s"like_$baseTable", s"select_$baseTable") {
+        val localTz = TimeZone.getDefault()
+        val localTzId = localTz.getID()
```
`localTz` and `localTzId` aren't used.

ueshin commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/ParquetHiveCompatibilitySuite.scala:
```diff
@@ -141,4 +160,326 @@ class ParquetHiveCompatibilitySuite extends ParquetCompatibilityTest with TestHi
+        spark.sql(
+          raw"""ALTER TABLE $baseTable SET TBLPROPERTIES ($key="America/Los_Angeles")""")
+        checkHasTz(baseTable, Some("America/Los_Angeles"))
+        spark.sql( raw"""ALTER TABLE $baseTable SET TBLPROPERTIES ($key="UTC")""")
```
nit: remove extra white space, and two more below this.

SparkQA commented:
**[Test build #75966 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75966/testReport)** for PR 16781 at commit [`44a8bbb`](https://github.com/apache/spark/commit/44a8bbb17484a61dd984fcb451e3b1be8c539e9f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/75966/
Test PASSed.

ueshin reviewed (commented):
@squito Thank you for working on this!
I checked updates and added some comments.

Btw, can you fix the partitioned tables?

squito commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/ParquetHiveCompatibilitySuite.scala:
```diff
@@ -42,6 +52,15 @@ class ParquetHiveCompatibilitySuite extends ParquetCompatibilityTest with TestHi
        """.stripMargin)
   }
 
+  override def afterEach(): Unit = {
```
I was probably just being a little paranoid, perhaps I had missed a `withTable` somewhere.  In the current code, things work just fine if I remove them.

squito commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/ParquetHiveCompatibilitySuite.scala:
```diff
@@ -141,4 +160,326 @@ class ParquetHiveCompatibilitySuite extends ParquetCompatibilityTest with TestHi
+    val schema = StructType(Seq(
+      StructField("display", StringType, true)
+    ))
+    val df = spark.createDataFrame(rowRdd, schema)
```
thanks, appreciate the help simplifying this.  I had a feeling it was more complex than it needed to be :)

squito committed:
review feedback

SparkQA commented:
**[Test build #76141 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76141/testReport)** for PR 16781 at commit [`e31657a`](https://github.com/apache/spark/commit/e31657a1b65ab73d52651c4e0b018d457e44d47f).

squito commented:
@ueshin updated per your feedback.

I should have explained that the last update *did* handle partition tables (it added the second call to `getStorageTzOptions` in `HiveMetastoreCatalog`), though I didn't have any tests for it.  It took me a while to figure out how to do it, but this update does include tests for creating partitioned tables and reading from them.  (the tests are becoming huge, but I think its worth testing all of the permutations.)

SparkQA commented:
**[Test build #76141 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76141/testReport)** for PR 16781 at commit [`e31657a`](https://github.com/apache/spark/commit/e31657a1b65ab73d52651c4e0b018d457e44d47f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76141/
Test PASSed.

ueshin commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/ParquetHiveCompatibilitySuite.scala:
```diff
@@ -397,13 +392,38 @@ class ParquetHiveCompatibilitySuite extends ParquetCompatibilityTest with TestHi
+                         |  display string,
+                         |  ts timestamp
+                         |)
+                         |PARTITIONED BY (id bigint)
```
We should test for the partitioned table like `PARTITIONED BY (ts timestamp)`?

squito committed:
Merge branch 'master' into SPARK-12297

squito committed:
add check for partitioned tables

squito committed:
fix typo

squito committed:
review feedback

SparkQA commented:
**[Test build #76391 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76391/testReport)** for PR 16781 at commit [`fc17a2e`](https://github.com/apache/spark/commit/fc17a2edc8e46dd95019f8ece972610d40f3ed2c).

squito commented:
@ueshin sorry it took me a while to figure out how a table partitioned by timestamps work (I didn't even realize that was possible, I don't think it is in hive?) and I was traveling.

The good news is that partitioning by timestamp works just fine.  Since the ts is stored as a string anyway, and [converted using the session tz already](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala#L135), it already works.  I added one minimal test on this -- when the partitioned table is written, the correct partition dirs are created regardless of the timezone combinations.

In particular, it doesn't make sense to do tests like the existing ones, where we write or read "unadjusted" data, bypassing the hive tables, and then make sure the right adjustments are applied when you perform the reverse action via the hive table; the partition values are correct whether you use the hive table & adjustment property or not.

Let me know if you think more tests are required.

SparkQA commented:
**[Test build #76391 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76391/testReport)** for PR 16781 at commit [`fc17a2e`](https://github.com/apache/spark/commit/fc17a2edc8e46dd95019f8ece972610d40f3ed2c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class ParquetHiveCompatibilitySuite extends ParquetCompatibilityTest with TestHiveSingleton `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76391/
Test PASSed.

ueshin commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/ParquetHiveCompatibilitySuite.scala:
```diff
@@ -141,4 +152,373 @@ class ParquetHiveCompatibilitySuite extends ParquetCompatibilityTest with TestHi
+        val defaultTz = None
+        // check that created tables have correct TBLPROPERTIES
+        val tblProperties = explicitTz.map {
+          tz => raw"""TBLPROPERTIES ($key="$tz")"""
```
Let's use `s""` instead of `raw""` if possible. And also elsewhere in the same way.

ueshin commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/ParquetHiveCompatibilitySuite.scala:
```diff
@@ -141,4 +152,373 @@ class ParquetHiveCompatibilitySuite extends ParquetCompatibilityTest with TestHi
+  val timestampTimezoneToMillis = {
+    val originalTz = TimeZone.getDefault
+    try {
+      (for {
```
Let's use `flatMap { .. map { ... } }`.

ueshin commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala:
```diff
@@ -29,6 +29,8 @@ import org.apache.spark.sql.catalyst.{QualifiedTableName, TableIdentifier}
 import org.apache.spark.sql.catalyst.plans.logical._
 import org.apache.spark.sql.execution.datasources._
+import org.apache.spark.sql.execution.datasources.parquet.ParquetFileFormat
+import org.apache.spark.sql.internal.SQLConf
```
nit: unnecessary import.

ueshin commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/ParquetHiveCompatibilitySuite.scala:
```diff
@@ -141,4 +152,373 @@ class ParquetHiveCompatibilitySuite extends ParquetCompatibilityTest with TestHi
+          tz => raw"""TBLPROPERTIES ($key="$tz")"""
+        }.getOrElse("")
+
+
```
nit: remove extra line.

ueshin commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/ParquetHiveCompatibilitySuite.scala:
```diff
@@ -141,4 +152,373 @@ class ParquetHiveCompatibilitySuite extends ParquetCompatibilityTest with TestHi
+      baseTable: String,
+      explicitTz: Option[String],
+      sessionTzOpt: Option[String]): Unit = {
+      val key = ParquetFileFormat.PARQUET_TIMEZONE_TABLE_PROPERTY
```
nit: indent

ueshin reviewed (commented):
@squito Thank you for working on this!
The behavior looks good for me.
I left some minor comments.
Thanks!

squito committed:
review feedback

SparkQA commented:
**[Test build #76419 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76419/testReport)** for PR 16781 at commit [`2537437`](https://github.com/apache/spark/commit/2537437a0c1f022acc215518e4728d24f6f6cf97).

SparkQA commented:
**[Test build #76419 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76419/testReport)** for PR 16781 at commit [`2537437`](https://github.com/apache/spark/commit/2537437a0c1f022acc215518e4728d24f6f6cf97).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76419/
Test PASSed.

ueshin commented:
Jenkins, retest this please.

SparkQA commented:
**[Test build #76556 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76556/testReport)** for PR 16781 at commit [`2537437`](https://github.com/apache/spark/commit/2537437a0c1f022acc215518e4728d24f6f6cf97).

ueshin commented:
LGTM, pending Jenkins.

SparkQA commented:
**[Test build #76556 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/76556/testReport)** for PR 16781 at commit [`2537437`](https://github.com/apache/spark/commit/2537437a0c1f022acc215518e4728d24f6f6cf97).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/76556/
Test PASSed.

ueshin commented:
Thanks! Merging to master.

squito commented:
great!  thanks @ueshin

rxin commented:
Did we conduct any performance tests on this patch?
